packages/server/modules/gatekeeper/services/workspacePlans.ts

```diff
@@ -0,0 +1,99 @@
+import { UpsertWorkspacePlan } from '@/modules/gatekeeper/domain/billing'
+import { InvalidWorkspacePlanStatus } from '@/modules/gatekeeper/errors/billing'
+import { WorkspacePlan } from '@/modules/gatekeeperCore/domain/billing'
+import { EventBusEmit } from '@/modules/shared/services/eventBus'
+import { GetWorkspace } from '@/modules/workspaces/domain/operations'
+import { WorkspaceNotFoundError } from '@/modules/workspaces/errors/workspace'
+import { throwUncoveredError } from '@speckle/shared'
+
+export const updateWorkspacePlanFactory =
+  ({
+    getWorkspace,
+    upsertWorkspacePlan,
+    emitEvent
+  }: {
+    getWorkspace: GetWorkspace
+    // im using the generic function here, cause the service is
+    // responsible for protecting the permutations
+    upsertWorkspacePlan: UpsertWorkspacePlan
+    emitEvent: EventBusEmit
+  }) =>
+  async ({
+    workspaceId,
+    name,
+    status
+  }: Pick<WorkspacePlan, 'workspaceId' | 'name' | 'status'>): Promise<void> => {
+    const workspace = await getWorkspace({
+      workspaceId
+    })
+    if (!workspace) throw new WorkspaceNotFoundError()
+    const createdAt = new Date()
+    switch (name) {
+      case 'starter':
+        switch (status) {
+          case 'trial':
+          case 'expired':
+            await upsertWorkspacePlan({
+              workspacePlan: { workspaceId, status, name, createdAt }
+            })
+            break
+          case 'valid':
+          case 'cancelationScheduled':
+          case 'canceled':
+          case 'paymentFailed':
+            await upsertWorkspacePlan({
+              workspacePlan: { workspaceId, status, name, createdAt }
```

Comment:
The entire `starter` block appears to be handled the same.

Reply:
ur right, this is a legacy, where the repo functions were specific for the plan type before